hbase-server/src/main/java/org/apache/hadoop/hbase/master/ActiveMasterManager.java

```diff
@@ -57,12 +56,18 @@
   final AtomicBoolean clusterHasActiveMaster = new AtomicBoolean(false);
   final AtomicBoolean clusterShutDown = new AtomicBoolean(false);
 
+  // This server's information.
   private final ServerName sn;
   private int infoPort;
   private final Server master;
 
+  // Active master's server name. Invalidated anytime active master changes (based on ZK
+  // notifications) and lazily fetched on-demand.
+  // ServerName is immutable, so we don't need heavy synchronization around it.
```

Comment:
Why is the `AtomicReference` necessary at all? If concurrent threads are updating this value, last write wins. The `ServerName` struct is assigned as a complete piece of data, only one field is updated, so there's no worry of seeing half of a write. If two threads find a `null` value, they'll both reach out to ZK to retrieve a value, so you're not preventing duplicate work. If the duplicate work thing isn't a concern, I think a `volatile` field will do the trick.